communication/src/coap_channel_new.cpp

```diff
@@ -0,0 +1,1457 @@
+
+namespace particle::protocol::experimental {
+
+namespace {
+
+const size_t MAX_TOKEN_SIZE = sizeof(token_t); // TODO: Support longer tokens
+const size_t MAX_TAG_SIZE = 8; // Maximum size of an ETag (RFC 7252) or Request-Tag (RFC 9175) option
+
+const unsigned BLOCK_SZX = 6; // Value of the SZX field for 1024-byte blocks (RFC 7959, 2.2)
+static_assert(COAP_BLOCK_SIZE == 1024); // When changing the block size, make sure to update BLOCK_SZX accordingly
+
+const unsigned DEFAULT_REQUEST_TIMEOUT = 60000;
+
+static_assert(COAP_INVALID_REQUEST_ID == 0); // Used by value in the code
+
+unsigned encodeBlockOption(int num, bool m) {
+    unsigned opt = (num << 4) | BLOCK_SZX;
+    if (m) {
+        opt |= 0x08;
+    }
+    return opt;
+}
+
+int decodeBlockOption(unsigned opt, int& num, bool& m) {
+    unsigned szx = opt & 0x07;
+    if (szx != BLOCK_SZX) {
+        // Server is required to use exactly the same block size
+        return SYSTEM_ERROR_NOT_SUPPORTED;
+    }
+    num = opt >> 4;
+    m = opt & 0x08;
+    return 0;
+}
+
+bool isValidCoapMethod(int method) {
+    switch (method) {
+    case COAP_METHOD_GET:
+    case COAP_METHOD_POST:
+    case COAP_METHOD_PUT:
+    case COAP_METHOD_DELETE:
+        return true;
+    default:
+        return false;
+    }
+}
+
+// TODO: Use a generic intrusive list container
+template<typename T, typename E, typename = std::enable_if_t<std::is_base_of_v<T, E>>>
+inline void addToList(T*& head, E* elem) {
+    assert(!elem->next && !elem->prev); // Debug-only
+    elem->prev = nullptr;
+    elem->next = head;
+    if (head) {
+        head->prev = elem;
+    }
+    head = elem;
+}
+
+template<typename T, typename E, typename = std::enable_if_t<std::is_base_of_v<T, E> || std::is_base_of_v<E, T>>>
+inline void removeFromList(T*& head, E* elem) {
+    assert(elem->next || elem->prev);
+    if (elem->prev) {
+        assert(head != elem);
+        elem->prev->next = elem->next;
+    } else {
+        assert(head == elem);
+        head = static_cast<T*>(elem->next);
+    }
+    if (elem->next) {
+        elem->next->prev = elem->prev;
+    }
+#ifndef NDEBUG
+    elem->next = nullptr;
+    elem->prev = nullptr;
+#endif
+}
+
+template<typename T, typename F>
+inline T* findInList(T* head, const F& fn) {
+    while (head) {
+        if (fn(head)) {
+            return head;
+        }
+        head = static_cast<T*>(head->next);
+    }
+    return nullptr;
+}
+
+template<typename T, typename F>
+inline void forEachInList(T* head, const F& fn) {
+    while (head) {
+        auto next = head->next; // In case the callback deletes the element
+        fn(head);
+        head = static_cast<T*>(next);
+    }
+}
+
+template<typename T, typename E>
+inline void addRefToList(T*& head, RefCountPtr<E> elem) {
+    addToList(head, elem.unwrap());
+}
+
+template<typename T, typename E>
+inline void removeRefFromList(T*& head, const RefCountPtr<E>& elem) {
+    removeFromList(head, elem.get());
+    elem->release();
```

Comment:
It looks odd that you have to manually release the `RefCount` here. Is there a way to make it symmeric with `addRefToList` by using `RefCountPtr::wrap` and letting the RAII class `RefCountPtr` call `release` on the inner pointer?

Reply:
Discussed [here](https://github.com/particle-iot/device-os/pull/2715#discussion_r1468230265).